tendermint/src/block/commit_sig.rs

```diff
@@ -1,79 +1,118 @@
+        /// Validator address
+        validator_address: account::Id,
+    },
+    */
+    // >>> end of real code
+    /// voted for the Commit.BlockID.
+    BlockIDFlagCommit {
+        /// Validator address
+        validator_address: account::Id,
+        /// Timestamp of vote
+        timestamp: Time,
+        /// Signature of vote
+        signature: Signature,
+    },
+    /// voted for nil.
+    BlockIDFlagNil {
+        /// Validator address
+        validator_address: account::Id,
+        /// Timestamp of vote
+        timestamp: Time,
+        /// Signature of vote
+        signature: Signature,
+    },
 }
 
-impl CommitSig {
-    /// Checks if a validator's vote is absent
-    pub fn is_absent(&self) -> bool {
-        self.block_id_flag == BlockIDFlag::BlockIDFlagAbsent
+impl TryFrom<RawCommitSig> for CommitSig {
+    type Error = &'static str;
+
+    fn try_from(value: RawCommitSig) -> Result<Self, Self::Error> {
+        // Validate CommitSig (strict)
+        match value.block_id_flag {
+            BlockIDFlag::BlockIDFlagAbsent => {
+                if value.timestamp.is_some() {
+                    // <<< Compatibility code for https://github.com/informalsystems/tendermint-rs/issues/259
+                    if value.timestamp.unwrap()
+                        != Time::parse_from_rfc3339("0001-01-01T00:00:00Z").unwrap()
+                    {
+                        return Err("timestamp is present for BlockIDFlagAbsent CommitSig");
+                    }
+                    // === Real code after compatibility issue is resolved
+                    /*
+                    return Err("timestamp is present for BlockIDFlagAbsent CommitSig");
+                    */
+                    // >>> end of real code
+                }
+                if value.signature.is_some() {
+                    return Err("signature is present for BlockIDFlagAbsent CommitSig");
+                }
+                // <<< Compatibility code for https://github.com/informalsystems/tendermint-rs/issues/260
+                Ok(CommitSig::BlockIDFlagAbsent)
+                // === Real code after compatibility issue is resolved
+                /*
+                Ok(CommitSig::BlockIDFlagAbsent {
+                    validator_address: value.validator_address,
+                })
+                */
+                // >>> end of real code
+            }
+            BlockIDFlag::BlockIDFlagCommit => {
+                if value.timestamp.is_none() {
+                    Err("timestamp is null for BlockIDFlagCommit CommitSig")
+                } else if value.signature.is_none() {
+                    Err("signature is null for BlockIDFlagCommit CommitSig")
+                } else {
+                    Ok(CommitSig::BlockIDFlagCommit {
+                        // <<< Compatibility code for https://github.com/informalsystems/tendermint-rs/issues/260
+                        validator_address: value.validator_address.unwrap(),
```

Comment:
Can't this unwrap panic here if `validator_address` is None? You could use `ok_or_else()` in combination with `?` to bubble up an error instead of potentially panicing.

Reply:
Yeah, I can do that - although since I'm already checking specifically for timestamp and signature, I could just check for validator_address too. (Giving a more relevant error message.)

If the Go code would be compatible with ADR-025, this wouldn't be a problem, because validator_address wouldn't be an Option<>. (See the "real code" below.)

Reply:
Panics removed.